utils/s2n_blob.c

```diff
@@ -43,9 +43,9 @@ int s2n_blob_init(struct s2n_blob *b, uint8_t * data, uint32_t size)
 
 int s2n_blob_zero(struct s2n_blob *b)
 {
+    PRECONDITION_POSIX(s2n_blob_is_valid(b));
     memset_check(b->data, 0, b->size);
```

Comment:
Should this be to `size` or `allocated`?

Reply:
That is actually probably a question for the s2n team. @colmmacc ?

Reply:
I didn't change the implementation, but it makes sense to me only zero the space that is actually used which is represented by size. We could change to allocated to be extra careful?

Reply:
@camshaft ?

Reply:
It's theoretically safer to zero the whole thing but after doing an audit on the codebase I think it'll break callers that don't set the `allocated` field. A few examples:

https://github.com/awslabs/s2n/blob/c8dd794088f2748d4840c85cc1d9e16232add806/tls/s2n_change_cipher_spec.c#L47
https://github.com/awslabs/s2n/blob/c8dd794088f2748d4840c85cc1d9e16232add806/tls/s2n_change_cipher_spec.c#L69
https://github.com/awslabs/s2n/blob/c8dd794088f2748d4840c85cc1d9e16232add806/tls/s2n_server_finished.c#L65
https://github.com/awslabs/s2n/blob/c8dd794088f2748d4840c85cc1d9e16232add806/tls/s2n_client_key_exchange.c#L82

Reply:
So if you were to change it to zero the whole thing, without breaking existing assumptions, you'd have to do something like:

```c
#define MAX(a, b, result) \
   do { \
      __typeof__ (a) _a = (a); \
      __typeof__ (b) _b = (b); \
      *(result) = _a > _b ? _a : _b; \
   } while(0)

size_t max = 0;
MAX(b->allocated, b->size, &max);
memset_check(b->data, 0, max);
```

Reply:
Here's another example with that assumption:

https://github.com/awslabs/s2n/blob/c8dd794088f2748d4840c85cc1d9e16232add806/utils/s2n_mem.c#L177-L179

`s2n_blob_slice` sets `allocated` to `0`:

https://github.com/awslabs/s2n/blob/c8dd794088f2748d4840c85cc1d9e16232add806/utils/s2n_blob.c#L60

Reply:
@camshaft I implemented as suggested. Could you take another look?